vllm/sequence.py

```diff
@@ -912,6 +941,26 @@ def prune(self,
             self.seq_ids = seq_ids
 
 
+class SamplingController:
```

Comment:
Make it an abstract class?

Reply:
Also in the docstring, we should probably mention this class is a singleton and stateful. And prepare & transform logits need to be called 1 tot 1

Reply:
Added docstrings.

It seems python doesn't have abstract classes, only abstract methods. I can imagine use cases where you only override some of these methods, and none of them you have to override, so I don't think it makes sense to make any of them abstract.

One thing we could do is to make the engine.sampling_controller field non-optional and use this base class as a no-op implementation. Not sure if that would be cleaner though?